fasta/src/main/java/org/pharmgkb/parsers/fasta/MultilineFastaSequenceParser.java

```diff
@@ -0,0 +1,210 @@
+public class MultilineFastaSequenceParser implements MultilineParser<FastaSequence> {
+
+	private final boolean m_allowComments;
+	private final boolean m_allowBlankLines;
+	private final String m_terminationString;
+
+	private String currentHeader = null;
+	private String currentSequence = null;
+	private boolean m_hitTerm = false;
+	private boolean m_hasTerm = false;
+
+	private static int m_nLines = 0;
+
+	public MultilineFastaSequenceParser(@Nonnull Builder builder) {
+		m_allowComments = builder.m_allowComments;
+		m_allowBlankLines = builder.m_allowBlankLines;
+		m_terminationString = builder.m_terminationString;
+	}
+
+	@Nonnull
+	public Runnable getCloseHandler() {
+		return () -> {
+			if (!m_hitTerm) throw new IllegalStateException("Stream never hit terminal sequence; appears not to have completed");
+		};
+	}
+
+	@Nonnull
+	private Stream<String> appendTermination(@Nonnull Stream<String> stream) {
+		m_hasTerm = true;
+		return Stream.concat(stream, Stream.of(m_terminationString));
+	}
+
+	@Nonnull
+	@Override
+	public Stream<FastaSequence> parseAll(@Nonnull Stream<String> stream) throws IOException, BadDataFormatException {
+		return appendTermination(stream).flatMap(this);
+	}
+
+	/**
+	 * <strong>Calling {@link #parseAll(Stream)} instead is strongly recommended.</strong>
+	 */
+	@Nonnull
+	@Override
+	public Stream<FastaSequence> apply(@Nonnull String line) {
+
+		if (!m_hasTerm) {
+			throw new IllegalStateException("Must call with parseAll or collectAll rather than apply");
+		}
+
+		m_nLines++;
+
+		if (line.equals(m_terminationString)) {
+			m_hitTerm = true;
+			if (currentHeader == null) { // happens if we read an empty source
+				return Stream.empty();
+			}
+			return Stream.of(new FastaSequence(currentHeader, currentSequence));
+		}
+
+		FastaSequence seq = readNext(line);
+		if (seq == null) return Stream.empty();
+		return Stream.of(seq);
+	}
+
+	@Nullable
+	private FastaSequence readNext(String line) {
+
+		if (line.startsWith(">")) {
+
+			FastaSequence seq = null;
+
+			// currentHeader is null only for the first header
+			// after reading each subsequent header, we know we've finished the sequence
+			// so, return that
+			if (currentHeader != null) {
+				try {
+					seq = new FastaSequence(currentHeader, currentSequence);
+				} catch (IllegalArgumentException e) {
+					throw new BadDataFormatException(e);
+				}
+			}
+
+			// reset
+			currentHeader = line.substring(1);
+			currentSequence = "";
+
+			return seq;
+
+		} else if (!m_allowBlankLines || !line.isEmpty()) { // don't trim
+			if (!m_allowComments || !line.startsWith(";")) {
+				if (currentHeader == null) {
+					throw new BadDataFormatException("Read sequence line \"" + line + "\" without header");
+				}
+				currentSequence += line;
+			}
+		}
+		return null;
+	}
+
+	@Override
+	public long nLinesProcessed() {
+		return m_nLines;
+	}
+
+	/**
+	 * Throws an {@link IllegalStateException} if the termination sequence of the stream was not reached.
```

Comment:
This is awesome documentation!  :)

I had a bunch of questions until I read it...

Note that without a few `<p>` tags this will all run together.

According to the javadoc on `finalize`:

> Any exception thrown by the finalize method causes the finalization of this object to be halted, but is otherwise ignored.

See https://stackoverflow.com/questions/18683996/object-finalize-method-ignoring-exceptions-thrown for a test of this.

So you won't actually get the notification you're expecting.  Given that, is there any point to this `finalize()` method if all you're looking for is a potential error?

Perhaps just log an error here and call it a day?